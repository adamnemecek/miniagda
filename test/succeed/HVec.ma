-- 2010-06-30 heterogeneous vectors
-- 2012-01-22 parameters gone from constructors

data Unit : Set { unit : Unit }

data Prod [i : Size] (A : Set i) (B : Set i) : Set i
--{ pair : A -> B -> Prod i A B
{ pair : (fst : A) -> (snd : B) -> Prod i A B
}

fun fst' : [i : Size] -> [A : Set i] -> [B : Set i] -> Prod i A B -> A
{ fst' i A B (pair a b) = a
}

data List [i : Size] (A : Set i) : Set i
{ nil  : List i A
; cons : A -> List i A -> List i A
}

-- recursive heterogeneous vectors
fun HVecR : List 1 Set -> Set
{ HVecR (nil) = Unit
; HVecR (cons A As) = Prod 0 A (HVecR As)
}

-- inductive heterogeneous vectors
data HVec : List 1 Set -> Set 1
{ vnil  : HVec (nil)
; vcons : [A : Set] -> [As : List 1 Set] ->
          A -> HVec As -> HVec (cons A As)
}
