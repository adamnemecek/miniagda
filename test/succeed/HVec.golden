MiniAgda by Andreas Abel and Karl Mehltretter
--- opening "HVec.ma" ---
--- scope checking ---
--- type checking ---
type  Unit : Set
term  Unit.unit : < Unit.unit : Unit >
ty-u  Prod : .[i : Size] -> ^(A : Set i) -> ^(B : Set i) -> Set i
term  Prod.pair : .[i : Size] -> .[A : Set i] -> .[B : Set i] -> ^(fst : A) -> ^(snd : B) -> < Prod.pair fst snd : Prod [i] A B >
mixk  fst : .[i : Size] -> .[A : Set i] -> .[B : Set i] -> (pair : Prod [i] A B) -> A
{ fst [i] [A] [B] (Prod.pair #fst #snd) = #fst
}
mixk  snd : .[i : Size] -> .[A : Set i] -> .[B : Set i] -> (pair : Prod [i] A B) -> B
{ snd [i] [A] [B] (Prod.pair #fst #snd) = #snd
}
mixk  fst' : .[i : Size] -> .[A : Set i] -> .[B : Set i] -> Prod [i] A B -> A
{ fst' [i] [A] [B] (Prod.pair a b) = a
}
ty-u  List : .[i : Size] -> ^(A : Set i) -> Set i
term  List.nil : .[i : Size] -> .[A : Set i] -> < List.nil : List [i] A >
term  List.cons : .[i : Size] -> .[A : Set i] -> ^(y0 : A) -> ^(y1 : List [i] A) -> < List.cons y0 y1 : List [i] A >
type  HVecR : List [1] Set -> Set
{ HVecR List.nil = Unit
; HVecR (List.cons A As) = Prod [0] A (HVecR As)
}
ty-u  HVec : ^ List [1] Set -> Set 1
term  HVec.vnil : < HVec.vnil : HVec List.nil >
term  HVec.vcons : .[A : Set] -> .[As : List [1] Set] -> ^(y2 : A) -> ^(y3 : HVec As) -> < HVec.vcons A As y2 y3 : HVec (List.cons A As) >
--- evaluating ---
--- closing "HVec.ma" ---
